### DISCUSSION TITLE
How should I best model this "friends" relationship in schema?

### DISCUSSION THREAD

**Discussion by chen-rn** [prisma/prisma#18380] [Q&A]
**A bit of context:** I'm building a pretty simple consumer social app where users can add other users. This is my current User model, with a self-referencing many-to-many relationship.

```
model User {
  id                  String          @id @default(uuid())
  phoneNumber         String          @unique
  friends             User[]          @relation("UserFriends")
  friendsRelation     User[]          @relation("UserFriends") //we can ignore this one more or less
}
```

**My challenge**: I'm looking to implement a few "relationship-esq" properties. For example: relationship type = family, or notification preferences = muted until tomorrow.

I'm thinking of adding a new Friend model, that is connected to the user, like following.

```
model User {
  id                  String          @id @default(uuid())
  phoneNumber         String          @unique
  Friend              Friend[]        @relation("Friend")
}
model Friend {
  id                     String @id @default(uuid())
  user                   User   @relation("Friend", fields: [userNumber], references: [phoneNumber])
  userNumber             String
  relationshipStatus     String
  notificationPreference String
}
```

I'm fairly new to Prisma and relationship databases as a whole, so I'm trying to see if this would be an ideal way to implement something like this! The other option of course is to add a "notification"/"relationships" property in each User but that seems off.

**Comment by ludralph** [comment]
Hi @chen-rn 👋 

Your proposed solution of creating a separate Friend model to store relationship status and notification preferences is a good approach. This will allow you to easily manage and query the relationship properties between users without cluttering the User model.

Here are a few suggestions to improve your model:

Instead of using userNumber to reference the user's phoneNumber in the Friend model, you can directly reference the User model using the user field. This will simplify your schema and allow for easier queries.

Consider using an enum for the relationshipStatus field to limit the possible values to a predefined set. This will ensure data consistency and make queries easier to write.

Similarly, you can use an enum for the notificationPreference field to limit the possible values.

Here's the updated schema:
```js
model User {
  id                  String          @id @default(uuid())
  phoneNumber         String          @unique
  friends             Friend[]        @relation("UserFriends")
}

model Friend {
  id                     String   @id @default(uuid())
  user                   User     @relation(fields: [userId], references: [id])
  userId                 String
  relationshipStatus     RelationshipStatus
  notificationPreference NotificationPreference
}

enum RelationshipStatus {
  FAMILY
  FRIEND
  ACQUAINTANCE
}

enum NotificationPreference {
  MUTED_UNTIL_TOMORROW
  MUTED_FOREVER
  NORMAL
}
```
If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by chen-rn** [reply]
This is fantastic!

In my case, friends are "mutual". Meaning that if you're my friend, I'm also your friend. It seems that I'll have to do two separate prisma.friend.create() requests, one for each user. Is it possible to do something where I create a _relationship_ model with both users, and it's automatically linked to both user's Friend property?

Additionally, my linter tells me that I have to add an opposite relationship field in Friend like as follows, is that correct?
```  user      User      @relation("UserFriends",fields: [userId], references: [id])```

**Comment by chen-rn** [comment]
I'm not sure if this is the right answer here.

It seems that I'll need to include both users in the Friend model. Or else the Friend model is _only_ tied to the owner. Not the friend. Which defeats the entire purpose right?

**Reply by anhtuanleee** [reply]
do u have a solution?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
